Integrity locks fixes

jvillafanez opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next version of ownCloud.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" if the PR still has open tasks
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
List of changes in this PR:
* Fix possible race condition in the `MemcacheLockingProvider` causing problems during the lock release
* Ensure `releaseLock` and `changeLock` methods of the `MemcacheLockingProvider` only happen if you have acquire a lock previously, otherwise log a warning and ignore the action.
* Added warning log in the `AbstractLockingProvider` to check whether we're releasing locks during shutdown. Locks should be released as early as possible, so the log should never happen.
* Refactor `OC_Files::get()` method to ensure the locks are properly released.

## Related Issue
https://github.com/owncloud/enterprise/issues/4319

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
1. Create a folder structure such as "/big share/batches/b1/file.jpg" with multiple big files in the "b1" folder
2. Share the "/big share" folder via link
3. Run the following script to download the shared folder by multiple clients at the same time (it should match the request made by the browser)
    ```
    #!/bin/bash
    for i in {1..50}
    do
      curl -k -o /tmp/down$i 'https://<server>:<port>/s/<shareToken>/download?path=%2Fbatches%2Fb1&files=' &
    done
    ```

No stray locks are reported. The locks are properly cleanup in redis
## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

@DeepDiver1975 I've taken a quick look at the `DBLockingProvider` and it seems might flood the logs. Locks are always released during shutdown, the new "stray lock removal" warning implemented here will complain a lot.

We shouldn't rely on auto-removing the lock during shutdown, and think about it as a safety measure. The warning is intended to rise awareness and fix the problem.
Note that we want to release the locks as early as possible to prevent locking exceptions in other clients if we aren't doing anything with the file.

jvillafanez committed:
Fix possible race conditions and wrong lock path

jvillafanez committed:
Refactor legacy/files::get method and add warning log for stray locks

The log should help to identify cases where the locks aren't cleared
properly (actively). Ideally, such log shouldn't appear.
The legacy/files::get method has been partially refactored to ensure the
locks are cleared when downloading a folder or several files. The case
where the user downloads just one file hasn't changed. This case is
currently triggering the stray lock warning.

jvillafanez committed:
Ensure locks are removed in the getSingleFile call in legacy/files.php

jvillafanez committed:
Ensure you can release or change a lock if you've acquire it first

jvillafanez commented:
Need to decide what to do with the DB implementation. It will cause noise with this PR as it is at the moment.

JanAckermann commented:
For me this PR doesn't fix the issue.

1. Uploade a 300MB ZIP File
2. Create a Public Share Link
3. Run the Command
```➜ #!/bin/bash
for i in {1..50}
do
  curl -k -o /tmp/down$i.zip 'https://owncloud.local/index.php/s/LJgjmsF9Z6xmSTt/download' &
done
```

Results in:
1. All downloaded ZIP Files are corrupted
2. Curl complains many times `[38]    27542 exit 23    curl -k -o /tmp/down$i.zip`
3. Log File
```
{"reqId":"978564a0-3db0-465f-99de-6593664e1bf1","level":4,"time":"2020-12-23T10:06:54+00:00","remoteAddr":"127.0.0.1","user":"--","app":"webdav","method":"PROPFIND","url":"\/remote.php\/dav\/files\/admin\/","message":"Exception: HTTP\/1.1 503 Error: Call to a member function getUID() on null: {\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\ServiceUnavailable\",\"Message\":\"Error: Call to a member function getUID() on null\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/WildcardEmitterTrait.php(89): {closure}(*** sensitive parameters replaced ***)\\n#1 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(456): Sabre\\\\DAV\\\\Server->emit()\\n#2 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(253): Sabre\\\\DAV\\\\Server->invokeMethod()\\n#3 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(321): Sabre\\\\DAV\\\\Server->start()\\n#4 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(70): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(168): handleException()\\n#6 {main}\",\"File\":\"\\\/home\\\/jan\\\/working\\\/core\\\/remote.php\",\"Line\":68}"}
{"reqId":"P21iQbHLqyRZcXXmAY1B","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"yILGs9uo8xEAN3PSKlmN","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"Kxy1vQJqnUKAcwE09ZVn","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"qNXMY06m752qvU2rMIxT","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"Vj6LHRQo3WMxPZ9vWNzR","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"sXRwttJHd2akRFctluiH","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"l9IjDj1SwKNyo1oyOyej","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"PhOtcgMhp7IKKIyoBGiv","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"n4xnL8qXMXRImXN5E7jp","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"3ftnLIwmEcHtstlcUwXK","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"3xhU4PRWiSuIEDYQMbXG","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"BTWPwohJR4oKtFH3Ygzv","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"HhNuUgUiCO9pqudzrXGY","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"vrNiNRcZucGpZP4eoTqt","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"lGFUxAbKKW469TFxldUL","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"H9mhLpmX8Km8tImKuRNf","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"7EmB7t9OBpnjXWmEYAvA","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"s84c6aUYxwTNo3MXOBEI","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"XSfUYXcPUWG2T7KKBPxa","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"xSib7zrEA7z6MMCM3HbG","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"lYskcm7hfuoAgGMP9bFS","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"cmGITXQ2KgihLZiSCs2u","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"h2x5LKxdcHPWXdualKI0","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"u2aqGqisvLhj1zQliXLU","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"GXi5CtrcwWRFjQAeAUos","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"UmXe0PwuroD0zife4UH2","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"q8wmIO26URBSW4qBb85b","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"eh13zLXgjbnNQxKSpZ8t","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"KhxySZ3Hga60dN2F70iY","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"eJfpybAED1OfO0kyqlIY","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"QL2sLGEcFQ6AGsyjgyvI","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"odPkHGBK4XCSGHAmgbRx","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"f9PhU0ye6vZdWXjy81rU","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"opnDAS9ZWNkEoofICI7X","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"xqzRCXCM67lnpEWfPsL0","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"AHHZTLpBkQO84pvcVrKX","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"3HW9y9yFnSVSRRvE5Rmd","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"gpWGOxHuudei8lXeX3wD","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"RfoV9k90GnEKq951BzET","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"0rO8QAfGqCH0Ugny2Qmm","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"kMos6HN5CAst4fK6ZEh3","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"2zwNPhNa7xHs07fZMaEU","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"i65B6QUqXm783uHTG78v","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"zqvLQcLtWCsJM04sAMTQ","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"eL1nukVrgMomb6iAfh6K","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"lW2gy5CD23xNYqrHfI4L","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"er54A6oaQar404Q2vL43","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"jlG1nBmOsA3hrVPuOFJT","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"GKBBnsJD6TEoHUOrqqLe","level":2,"time":"2020-12-23T10:06:55+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/index.php\/s\/LJgjmsF9Z6xmSTt\/download","message":"cleaning 1 stray shared locks for files\/31dc0b00a27f87e6348bf39c3c243628"}
{"reqId":"aa78e863-64b1-4ebe-9c94-79469c1a9940","level":4,"time":"2020-12-23T10:07:56+00:00","remoteAddr":"127.0.0.1","user":"--","app":"webdav","method":"PROPFIND","url":"\/remote.php\/dav\/files\/admin\/","message":"Exception: HTTP\/1.1 503 Error: Call to a member function getUID() on null: {\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\ServiceUnavailable\",\"Message\":\"Error: Call to a member function getUID() on null\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/WildcardEmitterTrait.php(89): {closure}(*** sensitive parameters replaced ***)\\n#1 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(456): Sabre\\\\DAV\\\\Server->emit()\\n#2 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(253): Sabre\\\\DAV\\\\Server->invokeMethod()\\n#3 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(321): Sabre\\\\DAV\\\\Server->start()\\n#4 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(70): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(168): handleException()\\n#6 {main}\",\"File\":\"\\\/home\\\/jan\\\/working\\\/core\\\/remote.php\",\"Line\":68}"}
{"reqId":"a01a0103-d649-458c-bb27-1f819a9f9f95","level":4,"time":"2020-12-23T10:08:58+00:00","remoteAddr":"127.0.0.1","user":"--","app":"webdav","method":"PROPFIND","url":"\/remote.php\/dav\/files\/admin\/","message":"Exception: HTTP\/1.1 503 Error: Call to a member function getUID() on null: {\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\ServiceUnavailable\",\"Message\":\"Error: Call to a member function getUID() on null\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/WildcardEmitterTrait.php(89): {closure}(*** sensitive parameters replaced ***)\\n#1 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(456): Sabre\\\\DAV\\\\Server->emit()\\n#2 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(253): Sabre\\\\DAV\\\\Server->invokeMethod()\\n#3 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(321): Sabre\\\\DAV\\\\Server->start()\\n#4 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(70): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(168): handleException()\\n#6 {main}\",\"File\":\"\\\/home\\\/jan\\\/working\\\/core\\\/remote.php\",\"Line\":68}"}
{"reqId":"cdb8758c-560e-4eb7-aa36-80608f31b2a6","level":4,"time":"2020-12-23T10:10:00+00:00","remoteAddr":"127.0.0.1","user":"--","app":"webdav","method":"PROPFIND","url":"\/remote.php\/dav\/files\/admin\/","message":"Exception: HTTP\/1.1 503 Error: Call to a member function getUID() on null: {\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\ServiceUnavailable\",\"Message\":\"Error: Call to a member function getUID() on null\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/WildcardEmitterTrait.php(89): {closure}(*** sensitive parameters replaced ***)\\n#1 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(456): Sabre\\\\DAV\\\\Server->emit()\\n#2 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(253): Sabre\\\\DAV\\\\Server->invokeMethod()\\n#3 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(321): Sabre\\\\DAV\\\\Server->start()\\n#4 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(70): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(168): handleException()\\n#6 {main}\",\"File\":\"\\\/home\\\/jan\\\/working\\\/core\\\/remote.php\",\"Line\":68}"}
{"reqId":"e1abbd89-64c9-47cf-93ea-bd295572c836","level":4,"time":"2020-12-23T10:11:02+00:00","remoteAddr":"127.0.0.1","user":"--","app":"webdav","method":"PROPFIND","url":"\/remote.php\/dav\/files\/admin\/","message":"Exception: HTTP\/1.1 503 Error: Call to a member function getUID() on null: {\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\ServiceUnavailable\",\"Message\":\"Error: Call to a member function getUID() on null\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/WildcardEmitterTrait.php(89): {closure}(*** sensitive parameters replaced ***)\\n#1 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(456): Sabre\\\\DAV\\\\Server->emit()\\n#2 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(253): Sabre\\\\DAV\\\\Server->invokeMethod()\\n#3 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(321): Sabre\\\\DAV\\\\Server->start()\\n#4 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(70): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(168): handleException()\\n#6 {main}\",\"File\":\"\\\/home\\\/jan\\\/working\\\/core\\\/remote.php\",\"Line\":68}"}
{"reqId":"f3fc32ee-2f0b-4dfd-9e14-62fdd0650b81","level":4,"time":"2020-12-23T10:12:04+00:00","remoteAddr":"127.0.0.1","user":"--","app":"webdav","method":"PROPFIND","url":"\/remote.php\/dav\/files\/admin\/","message":"Exception: HTTP\/1.1 503 Error: Call to a member function getUID() on null: {\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\ServiceUnavailable\",\"Message\":\"Error: Call to a member function getUID() on null\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/WildcardEmitterTrait.php(89): {closure}(*** sensitive parameters replaced ***)\\n#1 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(456): Sabre\\\\DAV\\\\Server->emit()\\n#2 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(253): Sabre\\\\DAV\\\\Server->invokeMethod()\\n#3 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(321): Sabre\\\\DAV\\\\Server->start()\\n#4 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(70): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(168): handleException()\\n#6 {main}\",\"File\":\"\\\/home\\\/jan\\\/working\\\/core\\\/remote.php\",\"Line\":68}"}
{"reqId":"1948cbcb-add5-450f-8737-649a3b3e93ff","level":4,"time":"2020-12-23T10:13:06+00:00","remoteAddr":"127.0.0.1","user":"--","app":"webdav","method":"PROPFIND","url":"\/remote.php\/dav\/files\/admin\/","message":"Exception: HTTP\/1.1 503 Error: Call to a member function getUID() on null: {\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\ServiceUnavailable\",\"Message\":\"Error: Call to a member function getUID() on null\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/WildcardEmitterTrait.php(89): {closure}(*** sensitive parameters replaced ***)\\n#1 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(456): Sabre\\\\DAV\\\\Server->emit()\\n#2 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(253): Sabre\\\\DAV\\\\Server->invokeMethod()\\n#3 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(321): Sabre\\\\DAV\\\\Server->start()\\n#4 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(70): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(168): handleException()\\n#6 {main}\",\"File\":\"\\\/home\\\/jan\\\/working\\\/core\\\/remote.php\",\"Line\":68}"}
{"reqId":"973e2d59-6e0b-4d1e-928b-77ca2a5e216e","level":4,"time":"2020-12-23T10:14:08+00:00","remoteAddr":"127.0.0.1","user":"--","app":"webdav","method":"PROPFIND","url":"\/remote.php\/dav\/files\/admin\/","message":"Exception: HTTP\/1.1 503 Error: Call to a member function getUID() on null: {\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\ServiceUnavailable\",\"Message\":\"Error: Call to a member function getUID() on null\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/WildcardEmitterTrait.php(89): {closure}(*** sensitive parameters replaced ***)\\n#1 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(456): Sabre\\\\DAV\\\\Server->emit()\\n#2 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(253): Sabre\\\\DAV\\\\Server->invokeMethod()\\n#3 \\\/home\\\/jan\\\/working\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(321): Sabre\\\\DAV\\\\Server->start()\\n#4 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(70): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/home\\\/jan\\\/working\\\/core\\\/remote.php(168): handleException()\\n#6 {main}\",\"File\":\"\\\/home\\\/jan\\\/working\\\/core\\\/remote.php\",\"Line\":68}"}
```

jvillafanez committed:
Fix wrong condition while changing the lock type

jvillafanez commented:
Question regarding the failing tests: what is the expected behaviour of the `changeLock` function if you don't have acquired the lock previously?

It seems the old code didn't take this into account. The tests are expecting a LockedException to be thrown, but it doesn't seems a good idea to throw that exception when there is no lock being used in that scenario.

The current code silently ignores the call (it just logs a warning), which is what the tests are complaining for. If this is the behaviour we want, we need to adjusts the tests, as well as add new ones.

jvillafanez committed:
throw exception instead of ignore to keep the previous behaviour

Trying to change a lock that we haven't acquired before will throw an
exception to keep the current behaviour. This scenario should be
considered as a coding error, and should be fixed by changing the code;
throwing an exception is harsh but it shouldn't happen under normal
conditions.

jvillafanez commented:
> Question regarding the failing tests: what is the expected behaviour of the changeLock function if you don't have acquired the lock previously?

Decided to throw an exception in that case. It will be treated as a coding error, so even if throwing an exception in that scenario is harsh, it's an scenario that shouldn't happen if the code does things the right way.

jvillafanez commented:
Waiting for decision on https://github.com/owncloud/enterprise/issues/4319#issuecomment-759560208

jvillafanez committed:
Fix stray lock warning while uploading a file

jvillafanez committed:
Fix tests

Parent::tearDown moved down to allow restoring the services first, and
remove the counter for the shared locks when the counter reaches 0 in
order to not log unneeded messages

jvillafanez commented:
Need to decide what to do with the logs in https://github.com/owncloud/core/pull/38229/files#diff-26e84810b47f83429f363bc771fc6aaec61635a66b3a35e31c619ce4732230b1 since they'll likely be quite noisy for the DB implementation at the moment.

jvillafanez committed:
Reduce log level to info to reduce the noise for DB implementation

jvillafanez commented:
Reduced the log level to "info" because it's expected the admins to set a log level of "warning" or "error".

micbar reviewed (approved):
LGTM

phil-davis commented:
@micbar this looks like it missed 10.7.0. Do we want it in 10.7.0 or?
